expertsystem/data.py

```diff
@@ -75,36 +78,114 @@ def magnitude(self) -> float:
         return self.__magnitude
 
     @property
-    def projection(self) -> float:
+    def projection(self) -> Optional[float]:
         return self.__projection
 
     def __hash__(self) -> int:
         return hash(repr(self))
 
 
-class Particle(NamedTuple):
-    """Immutable data container for particle info."""
+class QuantumState(NamedTuple):
+    """Collection of properties defining a quantum mechanical state.
+
+    Related to `.Particle`, but can contain more specific quantum state
+    information, such as `.Spin` projection.
+    """
 
-    name: str
-    pid: int
     charge: int
-    spin: float
-    mass: float
+    spin: Spin
+    isospin: Optional[Spin] = None
     strangeness: int = 0
     charmness: int = 0
     bottomness: int = 0
     topness: int = 0
     baryon_number: int = 0
-    electron_number: int = 0
-    muon_number: int = 0
-    tau_number: int = 0
-    width: Optional[float] = None
-    isospin: Optional[Spin] = None
+    electron_lepton_number: int = 0
+    muon_lepton_number: int = 0
+    tau_lepton_number: int = 0
     parity: Optional[Parity] = None
     c_parity: Optional[Parity] = None
     g_parity: Optional[Parity] = None
 
 
+class ComplexEnergyState:
+    """Pole in the complex energy plane, with quantum numbers."""
+
+    def __init__(self, energy: complex, state: QuantumState):
+        self.__energy = complex(energy)
+        self.state: QuantumState = state
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, ComplexEnergyState):
+            return (
+                self.complex_energy == other.complex_energy
+                and self.state == other.state
+            )
+        raise NotImplementedError
+
+    def __repr__(self) -> str:
+        return str(self.__energy)
+
+    @property
+    def complex_energy(self) -> complex:
+        return self.__energy
+
+    @property
+    def mass(self) -> float:
+        return self.__energy.real
+
+    @property
+    def width(self) -> float:
+        return self.__energy.imag
+
+
+class Particle(ComplexEnergyState):
+    """Immutable container of data defining a physical particle.
+
+    Can **only** contain info that the `PDG <http://pdg.lbl.gov/>`_ would list.
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        name: str,
+        pid: int,
+        mass: float,
+        state: QuantumState,
+        width: float = 0.0,
```

Comment:
Can you reorder those to group mass and width together

Reply:
Depends on https://github.com/ComPWA/expertsystem/pull/135#discussion_r463944145. Now, `QuantumState` has to be placed before because it requires an argument.

Reply:
True. For `QuantumState` I would leave spin as a required argument, since its hard to chose a reasonable default (`Particle`: Spin(0, None), `ComplexEnergyState`: Spin(0,0)). I'm starting to think, making a spin projection optional was a mistake. It just too error prone as the user can screw up setting the Spin values incorrectly... That means we would have to go back to the completely separate class for `Particle`. What to you think?